### PR TITLE
[ci] Add release github action

### PR DESCRIPTION
**Pull request by cyanglaz**
Add a Github action workflow for automatically publish plugins and tag release.
This action only runs on post-submit and will wait for all other ci jobs to pass before starting.

Fixes https://github.com/flutter/flutter/issues/81446

Manual integration test has been done in https://github.com/cyanglaz/melos_test/actions

**This PR should only land after all the GitHub Secretes are set.**

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[the auto-formatter]: https://github.com/flutter/plugins/blob/master/script/tool/README.md#format-code

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      # Github Actions don't support templates so it is hard to share this snippet with another action
+      # If we eventually need to use this in more workflow, we could create a shell script that contains this
+      # snippet.
+      run: |
+        cd $HOME
+        git clone https://github.com/flutter/flutter.git --depth 1 -b stable _flutter
+        echo "$HOME/_flutter/bin" >> $GITHUB_PATH
+        echo "$HOME/.pub-cache/bin" >> $GITHUB_PATH
+        echo "$HOME/_flutter/.pub-cache/bin" >> $GITHUB_PATH
+        echo "$HOME/_flutter/bin/cache/dart-sdk/bin" >> $GITHUB_PATH
```
Comment (stuartmorgan) [inline .github/workflows/release.yml:22]: Why are these last three lines needed? `dart` is in `flutter/bin/` these days.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        cd $GITHUB_WORKSPACE
+    # Checks out a copy of the repo.
+    - name: Checkout code
```
Comment (stuartmorgan) [inline .github/workflows/release.yml:25]: Check out code

(checkout is a nout)

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f
+      with:
+        fetch-depth: 0 # Fetch all history so the tool can get all the tag to determine version.
```
Comment (stuartmorgan) [inline .github/workflows/release.yml:28]: s/tag/tags/

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        running-workflow-name: 'release'
+        repo-token: ${{ secrets.GITHUB_TOKEN }}
+        wait-interval: 10 # seconds
+        allowed-conclusions: success
+
+    - name: run release
+      run: |
+        git config --global user.name ${{ secrets.USER_NAME }}
+        git config --global user.email ${{ secrets.USER_EMAIL }}
+        dart ./script/tool/lib/src/main.dart publish-plugin --all-changed --base-sha=HEAD~ --skip-confirmation --remote=origin
```
Comment (stuartmorgan) [inline .github/workflows/release.yml:47]: How is the tag push authorization going to work? Do GA actions automatically have write access to the repo?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml (hunk 4 above)
Comment (cyanglaz) [reply .github/workflows/release.yml:47]: Yes, I think so.

File: .github/workflows/release.yml (hunk 1 above)
Comment (cyanglaz) [reply .github/workflows/release.yml:22]: Oh ill remove them.

**Comment by cyanglaz** [comment]
@stuartmorgan Fixed, PTAL. The secrets are set up as well. So this is ready to land once LGTM'd

**Comment by cyanglaz** [comment]
Friendly ping :) @stuartmorgan

**Review by stuartmorgan — APPROVED** [review]
LGTM!

**Pull request merged**